tooling/doc_generator/src/getters.rs

```diff
@@ -0,0 +1,503 @@
+        _ => {
+            let mut res = String::new();
+
+            let mut doc_find = true;
+            let mut iter = 2;
+            while doc_find && ((index as i32) - (iter as i32)) >= 0 {
+                match &tokens[index - iter] {
+                    Token::LineComment(doc, Some(DocStyle::Outer))
+                    | Token::BlockComment(doc, Some(DocStyle::Outer)) => {
+                        res.insert_str(0, &doc.to_string());
+                        iter += 1;
+                    }
+                    Token::Keyword(Keyword::Fn)
+                    | Token::Keyword(Keyword::Mod)
+                    | Token::Keyword(Keyword::Struct)
+                    | Token::Keyword(Keyword::Trait)
+                    | Token::Keyword(Keyword::Impl) => {
+                        doc_find = false;
+                    }
+                    _ => {
+                        iter += 1;
+                    }
+                }
+            }
+            res
+        }
+    };
+    res
+}
+
+/// Extracts an outer documentation comment associated with a code element from a list of tokens.
+
+/// The `outer_doc` function is used to extract an outer documentation comment associated with a
+/// code element from a list of tokens. Outer documentation comments are often used to provide
+/// high-level explanations or descriptions for the code element. The function starts at the
+/// specified index and searches for an outer documentation comment, returning the comment as a string
+/// along with the updated index.
+pub(crate) fn outer_doc(tokens: &[Token], index: usize) -> (String, usize) {
+    let mut i = index;
+    let mut res = tokens[i].to_string();
+    let mut doc_find = true;
+    while doc_find {
+        match &tokens[i + 1] {
+            Token::LineComment(doc, Some(DocStyle::Inner))
+            | Token::BlockComment(doc, Some(DocStyle::Inner)) => {
+                res.push_str(doc);
+                i += 1;
+            }
+            _ => {
+                doc_find = false;
+            }
+        }
+    }
+
+    if let Some(pos) = res.find(' ') {
+        res = res.split_off(pos + 1);
+    } else {
+        res.clear();
+    }
+
+    (res, i)
+}
+
+/// Reads and tokenizes the content of a source file, returning a vector of spanned tokens.
+
+/// The `get_doc` function reads the content of a source file specified by the `input_file` path,
+/// tokenizes the content, and returns the resulting vector of spanned tokens. This function is
+/// typically used for processing source code and extracting tokens for further analysis or documentation.
+pub(crate) fn get_doc(input_file: &str) -> Result<Vec<SpannedToken>, Box<dyn std::error::Error>> {
+    let mut file = File::open(input_file)?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+
+    let mut lexer = Lexer::new(&contents);
+    lexer = lexer.skip_comments(false);
+
+    let token = lexer.into_iter().map(|a| a.unwrap()).collect::<Vec<_>>();
```

Comment:
By the way, Lexer already implements the Iterator trait, so there is no need to call into_iter. And it's unlikely that we want to panic here.

Reply:
```rust
let comments =
    Lexer::new(source).skip_comments(false).collect::<Result<Vec<_>>>()?;
```